Connexion does not validate connexion.problem responses properly

### Description

Connexion does not validate `connexion.problem` responses properly. This is a regression introduced by 9bef8c46a699719f2616560fcdde7add2299d5af, and first appearing in version 2.1.0.

### Steps to reproduce

`spec.yml`:
```yaml
openapi: "3.0.0"
info:
  title: Sample API
  version: "0.1"
paths:
  /:
    get:
      operationId: main.get
      responses:
        403:
          description: Forbidden
          content:
            application/problem+json:
              schema:
                type: object
```

`main.py`:
```python
import connexion
import logging
import unittest

def get():
  return connexion.problem(403, "Forbidden", "Forbidden")

class Test(unittest.TestCase):
  def test(self):
    app = connexion.FlaskApp(
        __name__,
        specification_dir=".",
        options={ "swagger_ui": False })
    app.add_api("spec.yml", validate_responses=True)
    client = app.app.test_client()
    resp = client.get("/")
    self.assertEqual(resp.status_code, 403)

if __name__ == "__main__":
  logging.basicConfig(level=logging.INFO)
  unittest.main()
```

### Expected behaviour

```console
$ python main.py
.
----------------------------------------------------------------------
Ran 1 test in 0.017s

OK
```

### Actual behaviour

```console
$ python main.py
ERROR:connexion.decorators.validation:http://localhost/ validation error: None is not of type 'object'

Failed validating 'type' in schema:
    {'type': 'object'}

On instance:
    None
F
======================================================================
FAIL: test (__main__.Test)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gasmith/cx/main.py", line 21, in test
    self.assertEqual(resp.status_code, 403)
AssertionError: 500 != 403

----------------------------------------------------------------------
Ran 1 test in 0.023s

FAILED (failures=1)
```

### Bisection

`bisect.sh`:
```bash
export PYTHONPATH=$(pwd)
git checkout master
git bisect start
git bisect bad
git bisect good 1.5.3
git bisect run python main.py
```

Bisection output:
```
$ ./bisect.sh
...
9bef8c46a699719f2616560fcdde7add2299d5af is the first bad commit
commit 9bef8c46a699719f2616560fcdde7add2299d5af
Author: Diogo Dutra <diogodutradamata@gmail.com>
Date:   Tue Dec 11 05:28:58 2018 -0200

    Removed duplicated calls of get_response method of the Api classes (#809)

    On Operation decorators lifecycle the api.get_response method were always called twice.
    Because the response validation, we need to cast the response to the ConnexionResponse type.
    But the response casting could be made on the response validation decorator, not always. So, changes were made to cast the response correctly.
    The BeginOfRequestLifecycleDecorator class was removed because it only do the second call of api.get_response method. This second call only occurs if the response validation is needed.
    For a more logical naming, the EndOfRequestLifecycleDecorator class was renamed to RequestResponseDecorator.
    Fixes #664

:040000 040000 e68b6742cb39d7c4bb0a0349ef8cf63e16303ef3 0b0e738c93803db751213a1b946347665998dd79 M      connexion
bisect run success
```

### Additional info:

The same behavior is observed for both Python 2.7 and Python 3.5.

--- a/connexion/__init__.py
+++ b/connexion/__init__.py
@@ -7,8 +7,6 @@ REST API in as much detail as you want; then Connexion guarantees that it will w
 specified.
 """
 
-import werkzeug.exceptions as exceptions  # NOQA
-
 from .apps import AbstractApp  # NOQA
 from .apps.asynchronous import AsyncApp
 from .datastructures import NoContent  # NOQA

--- a/connexion/apps/flask.py
+++ b/connexion/apps/flask.py
@@ -14,7 +14,7 @@ from starlette.types import Receive, Scope, Send
 
 from connexion.apps.abstract import AbstractApp
 from connexion.decorators import FlaskDecorator
-from connexion.exceptions import ProblemException, ResolverError
+from connexion.exceptions import InternalServerError, ProblemException, ResolverError
 from connexion.frameworks import flask as flask_utils
 from connexion.jsonifier import Jsonifier
 from connexion.middleware.abstract import AbstractRoutingAPI, SpecMiddleware
@@ -127,18 +127,10 @@ class FlaskMiddlewareApp(SpecMiddleware):
     def common_error_handler(self, exception: Exception) -> FlaskResponse:
         """Default error handler."""
         if isinstance(exception, ProblemException):
-            response = problem(
-                status=exception.status,
-                title=exception.title,
-                detail=exception.detail,
-                type=exception.type,
-                instance=exception.instance,
-                headers=exception.headers,
-                ext=exception.ext,
-            )
+            response = exception.to_problem()
         else:
             if not isinstance(exception, werkzeug.exceptions.HTTPException):
-                exception = werkzeug.exceptions.InternalServerError()
+                exception = InternalServerError()
 
             response = problem(
                 title=exception.name,

--- a/connexion/exceptions.py
+++ b/connexion/exceptions.py
@@ -2,7 +2,7 @@
 This module defines Exception classes used by Connexion to generate a proper response.
 """
 
-import warnings
+import typing as t
 
 from jsonschema.exceptions import ValidationError
 from starlette.exceptions import HTTPException
@@ -14,10 +14,33 @@ class ConnexionException(Exception):
     pass
 
 
-class ProblemException(ConnexionException):
+class ResolverError(LookupError, ConnexionException):
+    def __init__(self, detail: str = "Unknown reason", *, exc_info: tuple) -> None:
+        """
+        :param detail: Reason why the resolver failed.
+        :param exc_info: If specified, gives details of the original exception
+            as returned by sys.exc_info()
+        """
+        self.reason = detail
+        self.exc_info = exc_info
+
+
+class InvalidSpecification(ValidationError, ConnexionException):
+    pass
+
+
+class MissingMiddleware(ConnexionException):
+    pass
+
+
+# HTTP ERRORS
+
+
+class ProblemException(HTTPException, ConnexionException):
     def __init__(
         self,
-        status=400,
+        *,
+        status=500,
         title=None,
         detail=None,
         type=None,
@@ -29,7 +52,7 @@ class ProblemException(ConnexionException):
         This exception holds arguments that are going to be passed to the
         `connexion.problem` function to generate a proper response.
         """
-        self.status = status
+        self.status = self.status_code = status
         self.title = title
         self.detail = detail
         self.type = type
@@ -38,11 +61,6 @@ class ProblemException(ConnexionException):
         self.ext = ext
 
     def to_problem(self):
-        warnings.warn(
-            "'to_problem' is planned to be removed in a future release. "
-            "Call connexion.problem.problem(..) instead to maintain the existing error response.",
-            DeprecationWarning,
-        )
         return problem(
             status=self.status,
             title=self.title,
@@ -54,96 +72,33 @@ class ProblemException(ConnexionException):
         )
 
 
-class ResolverError(LookupError):
-    def __init__(self, reason="Unknown reason", exc_info=None):
-        """
-        :param reason: Reason why the resolver failed.
-        :type reason: str
-        :param exc_info: If specified, gives details of the original exception
-            as returned by sys.exc_info()
-        :type exc_info: tuple | None
-        """
-        self.reason = reason
-        self.exc_info = exc_info
-
-    def __str__(self):  # pragma: no cover
-        return f"<ResolverError: {self.reason}>"
-
-    def __repr__(self):  # pragma: no cover
-        return f"<ResolverError: {self.reason}>"
+# CLIENT ERRORS (4XX)
 
 
-class InvalidSpecification(ConnexionException, ValidationError):
-    pass
-
-
-class MissingMiddleware(ConnexionException):
-    pass
-
-
-class NonConformingResponse(ProblemException):
-    def __init__(self, reason="Unknown Reason", message=None):
-        """
-        :param reason: Reason why the response did not conform to the specification
-        :type reason: str
-        """
-        super().__init__(status=500, title=reason, detail=message)
-        self.reason = reason
-        self.message = message
-
-    def __str__(self):  # pragma: no cover
-        return f"<NonConformingResponse: {self.reason}>"
-
-    def __repr__(self):  # pragma: no cover
-        return f"<NonConformingResponse: {self.reason}>"
-
-
-class AuthenticationProblem(ProblemException):
-    def __init__(self, status, title, detail):
+class ClientError(ProblemException):
+    def __init__(self, status: int = 400, title: str = None, *, detail: str = None):
         super().__init__(status=status, title=title, detail=detail)
 
 
-class ResolverProblem(ProblemException):
-    def __init__(self, status, title, detail):
-        super().__init__(status=status, title=title, detail=detail)
+class BadRequestProblem(ClientError):
+    def __init__(self, detail=None):
+        super().__init__(status=400, title="Bad Request", detail=detail)
 
 
-class BadRequestProblem(ProblemException):
-    def __init__(self, title="Bad Request", detail=None):
-        super().__init__(status=400, title=title, detail=detail)
+class ExtraParameterProblem(BadRequestProblem):
+    def __init__(self, *, param_type: str, extra_params: t.Iterable[str]):
+        detail = f"Extra {param_type} parameter(s) {','.join(extra_params)} not in spec"
+        super().__init__(detail=detail)
 
 
-class NotFoundProblem(ProblemException):
-
-    description = (
-        "The requested URL was not found on the server. If you entered the URL manually please "
-        "check your spelling and try again."
-    )
-
-    def __init__(self, title="Not Found", detail=description):
-        super().__init__(status=404, title=title, detail=detail)
+class TypeValidationError(BadRequestProblem):
+    def __init__(self, schema_type: str, parameter_type: str, parameter_name: str):
+        """Exception raised when type validation fails"""
+        detail = f"Wrong type, expected '{schema_type}' for {parameter_type} parameter '{parameter_name}'"
+        super().__init__(detail=detail)
 
 
-class UnsupportedMediaTypeProblem(ProblemException):
-    def __init__(self, title="Unsupported Media Type", detail=None):
-        super().__init__(status=415, title=title, detail=detail)
-
-
-class NonConformingResponseBody(NonConformingResponse):
-    def __init__(
-        self, message, reason="Response body does not conform to specification"
-    ):
-        super().__init__(reason=reason, message=message)
-
-
-class NonConformingResponseHeaders(NonConformingResponse):
-    def __init__(
-        self, message, reason="Response headers do not conform to specification"
-    ):
-        super().__init__(reason=reason, message=message)
-
-
-class Unauthorized(HTTPException):
+class Unauthorized(ClientError):
 
     description = (
         "The server could not verify that you are authorized to access"
@@ -152,8 +107,8 @@ class Unauthorized(HTTPException):
         " how to supply the credentials required."
     )
 
-    def __init__(self, detail: str = description, **kwargs):
-        super().__init__(401, detail=detail, **kwargs)
+    def __init__(self, detail: str = description):
+        super().__init__(401, title="Unauthorized", detail=detail)
 
 
 class OAuthProblem(Unauthorized):
@@ -161,67 +116,77 @@ class OAuthProblem(Unauthorized):
 
 
 class OAuthResponseProblem(OAuthProblem):
-    def __init__(self, token_response, **kwargs):
-        self.token_response = token_response
-        super().__init__(**kwargs)
+    pass
 
 
 class Forbidden(HTTPException):
-
-    description = (
-        "You don't have the permission to access the requested"
-        " resource. It is either read-protected or not readable by the"
-        " server."
-    )
-
-    def __init__(self, detail: str = description, **kwargs):
-        super().__init__(403, detail=detail, **kwargs)
+    def __init__(self, detail: t.Optional[str] = None):
+        if detail is None:
+            detail = (
+                "You don't have the permission to access the requested"
+                " resource. It is either read-protected or not readable by the"
+                " server."
+            )
+        super().__init__(403, detail=detail)
 
 
 class OAuthScopeProblem(Forbidden):
-    def __init__(self, token_scopes, required_scopes, **kwargs):
-        self.required_scopes = required_scopes
-        self.token_scopes = token_scopes
-
-        super().__init__(**kwargs)
+    pass
 
 
-class ExtraParameterProblem(ProblemException):
+class UnsupportedMediaTypeProblem(ClientError):
+    def __init__(self, detail: t.Optional[str] = None):
+        super().__init__(status=415, title="Unsupported Media Type", detail=detail)
+
+
+# SERVER ERRORS (5XX)
+
+
+class ServerError(ProblemException):
     def __init__(
-        self, formdata_parameters, query_parameters, title=None, detail=None, **kwargs
+        self,
+        status: int = 500,
+        title: t.Optional[str] = None,
+        *,
+        detail: t.Optional[str] = None,
     ):
-        self.extra_formdata = formdata_parameters
-        self.extra_query = query_parameters
+        if title is None:
+            title = "Internal Server Error"
 
-        # This keep backwards compatibility with the old returns
+        super().__init__(status=status, title=title, detail=detail)
+
+
+class InternalServerError(ServerError):
+    def __init__(self, detail: t.Optional[str] = None):
         if detail is None:
-            if self.extra_query:
-                detail = "Extra {parameter_type} parameter(s) {extra_params} not in spec".format(
-                    parameter_type="query", extra_params=", ".join(self.extra_query)
-                )
-            elif self.extra_formdata:
-                detail = "Extra {parameter_type} parameter(s) {extra_params} not in spec".format(
-                    parameter_type="formData",
-                    extra_params=", ".join(self.extra_formdata),
-                )
-
-        super().__init__(title=title, detail=detail, **kwargs)
+            detail = (
+                "The server encountered an internal error and was unable to complete your "
+                "request. Either the server is overloaded or there is an error in the application."
+            )
+        super().__init__(status=500, title="Internal Server Error", detail=detail)
 
 
-class TypeValidationError(Exception):
-    def __init__(self, schema_type, parameter_type, parameter_name):
-        """
-        Exception raise when type validation fails
+class NonConformingResponse(InternalServerError):
+    def __init__(self, detail: t.Optional[str] = None):
+        super().__init__(detail=detail)
 
-        :type schema_type: str
-        :type parameter_type: str
-        :type parameter_name: str
-        :return:
-        """
-        self.schema_type = schema_type
-        self.parameter_type = parameter_type
-        self.parameter_name = parameter_name
 
-    def __str__(self):
-        msg = "Wrong type, expected '{schema_type}' for {parameter_type} parameter '{parameter_name}'"
-        return msg.format(**vars(self))
+class NonConformingResponseBody(NonConformingResponse):
+    def __init__(self, detail: t.Optional[str] = None):
+        if detail is None:
+            detail = "Response body does not conform to specification"
+
+        super().__init__(detail=detail)
+
+
+class NonConformingResponseHeaders(NonConformingResponse):
+    def __init__(self, detail: t.Optional[str] = None):
+        if detail is None:
+            detail = "Response headers do not conform to specification"
+
+        super().__init__(detail=detail)
+
+
+class ResolverProblem(ServerError):
+    def __init__(self, status: int = 501, *, detail: t.Optional[str] = None):
+        super().__init__(status=status, title="Not Implemented", detail=detail)

--- a/connexion/exceptions.py
+++ b/connexion/exceptions.py
@@ -124,7 +124,14 @@ class Forbidden(HTTPException):
 
 
 class OAuthScopeProblem(Forbidden):
-    pass
+    def __init__(self, token_scopes: list, required_scopes: list) -> None:
+        self.required_scopes = required_scopes
+        self.token_scopes = token_scopes
+        detail = (
+            f"Provided token does not have the required scopes. "
+            f"Provided: {token_scopes}; Required: {required_scopes}"
+        )
+        super().__init__(detail=detail)
 
 
 class UnsupportedMediaTypeProblem(ClientError):

--- a/connexion/exceptions.py
+++ b/connexion/exceptions.py
@@ -15,14 +15,7 @@ class ConnexionException(Exception):
 
 
 class ResolverError(LookupError, ConnexionException):
-    def __init__(self, detail: str = "Unknown reason", *, exc_info: tuple) -> None:
-        """
-        :param detail: Reason why the resolver failed.
-        :param exc_info: If specified, gives details of the original exception
-            as returned by sys.exc_info()
-        """
-        self.reason = detail
-        self.exc_info = exc_info
+    pass
 
 
 class InvalidSpecification(ValidationError, ConnexionException):

--- a/connexion/handlers.py
+++ b/connexion/handlers.py
@@ -26,7 +26,6 @@ class ResolverErrorHandler:
 
     def handle(self, *args, **kwargs):
         raise ResolverProblem(
-            title="Not Implemented",
             detail=self.exception.reason,
             status=self.status_code,
         )
@@ -55,7 +54,6 @@ class ResolverErrorHandler:
 
     async def __call__(self, *args, **kwargs):
         raise ResolverProblem(
-            title="Not Implemented",
             detail=self.exception.reason,
             status=self.status_code,
         )

--- a/connexion/handlers.py
+++ b/connexion/handlers.py
@@ -26,7 +26,7 @@ class ResolverErrorHandler:
 
     def handle(self, *args, **kwargs):
         raise ResolverProblem(
-            detail=self.exception.reason,
+            detail=self.exception.args[0],
             status=self.status_code,
         )
 
@@ -54,6 +54,6 @@ class ResolverErrorHandler:
 
     async def __call__(self, *args, **kwargs):
         raise ResolverProblem(
-            detail=self.exception.reason,
+            detail=self.exception.args[0],
             status=self.status_code,
         )

--- a/connexion/middleware/abstract.py
+++ b/connexion/middleware/abstract.py
@@ -1,7 +1,6 @@
 import abc
 import logging
 import pathlib
-import sys
 import typing as t
 
 from starlette.types import ASGIApp, Receive, Scope, Send
@@ -106,10 +105,10 @@ class AbstractRoutingAPI(AbstractSpecAPI, t.Generic[OP]):
                     if self.resolver_error_handler is not None:
                         self._add_resolver_error_handler(method, path, err)
                     else:
-                        self._handle_add_operation_error(path, method, err.exc_info)
-                except Exception:
+                        self._handle_add_operation_error(path, method, err)
+                except Exception as e:
                     # All other relevant exceptions should be handled as well.
-                    self._handle_add_operation_error(path, method, sys.exc_info())
+                    self._handle_add_operation_error(path, method, e)
 
     def add_operation(self, path: str, method: str) -> None:
         """
@@ -169,15 +168,12 @@ class AbstractRoutingAPI(AbstractSpecAPI, t.Generic[OP]):
         self._add_operation_internal(method, path, operation)
 
     def _handle_add_operation_error(
-        self, path: str, method: str, exc_info: tuple
+        self, path: str, method: str, exc: Exception
     ) -> None:
         url = f"{self.base_path}{path}"
-        error_msg = "Failed to add operation for {method} {url}".format(
-            method=method.upper(), url=url
-        )
+        error_msg = f"Failed to add operation for {method.upper()} {url}"
         logger.error(error_msg)
-        _type, value, traceback = exc_info
-        raise value.with_traceback(traceback)
+        raise exc from None
 
     def json_loads(self, data):
         return self.jsonifier.loads(data)

--- a/connexion/middleware/exceptions.py
+++ b/connexion/middleware/exceptions.py
@@ -1,13 +1,12 @@
 import json
 
-import werkzeug.exceptions
 from starlette.exceptions import ExceptionMiddleware as StarletteExceptionMiddleware
 from starlette.exceptions import HTTPException
 from starlette.requests import Request as StarletteRequest
 from starlette.responses import Response
 from starlette.types import Receive, Scope, Send
 
-from connexion.exceptions import ProblemException, problem
+from connexion.exceptions import InternalServerError, ProblemException, problem
 
 
 class ExceptionMiddleware(StarletteExceptionMiddleware):
@@ -19,27 +18,19 @@ class ExceptionMiddleware(StarletteExceptionMiddleware):
         self.add_exception_handler(ProblemException, self.problem_handler)
         self.add_exception_handler(Exception, self.common_error_handler)
 
-    def problem_handler(self, _request: StarletteRequest, exception: ProblemException):
-        connexion_response = problem(
-            status=exception.status,
-            title=exception.title,
-            detail=exception.detail,
-            type=exception.type,
-            instance=exception.instance,
-            headers=exception.headers,
-            ext=exception.ext,
-        )
+    @staticmethod
+    def problem_handler(_request: StarletteRequest, exception: ProblemException):
+        response = exception.to_problem()
 
         return Response(
-            content=json.dumps(connexion_response.body),
-            status_code=connexion_response.status_code,
-            media_type=connexion_response.mimetype,
-            headers=connexion_response.headers,
+            content=json.dumps(response.body),
+            status_code=response.status_code,
+            media_type=response.mimetype,
+            headers=response.headers,
         )
 
-    def http_exception(
-        self, _request: StarletteRequest, exc: HTTPException
-    ) -> Response:
+    @staticmethod
+    def http_exception(_request: StarletteRequest, exc: HTTPException) -> Response:
         headers = exc.headers
 
         connexion_response = problem(
@@ -53,16 +44,9 @@ class ExceptionMiddleware(StarletteExceptionMiddleware):
             headers=connexion_response.headers,
         )
 
-    def common_error_handler(
-        self, _request: StarletteRequest, exc: HTTPException
-    ) -> Response:
-        exception = werkzeug.exceptions.InternalServerError()
-
-        response = problem(
-            title=exception.name,
-            detail=exception.description,
-            status=exception.code,
-        )
+    @staticmethod
+    def common_error_handler(_request: StarletteRequest, exc: Exception) -> Response:
+        response = InternalServerError().to_problem()
 
         return Response(
             content=json.dumps(response.body),

--- a/connexion/middleware/exceptions.py
+++ b/connexion/middleware/exceptions.py
@@ -1,5 +1,3 @@
-import json
-
 from starlette.exceptions import ExceptionMiddleware as StarletteExceptionMiddleware
 from starlette.exceptions import HTTPException
 from starlette.requests import Request as StarletteRequest
@@ -23,7 +21,7 @@ class ExceptionMiddleware(StarletteExceptionMiddleware):
         response = exception.to_problem()
 
         return Response(
-            content=json.dumps(response.body),
+            content=response.body,
             status_code=response.status_code,
             media_type=response.mimetype,
             headers=response.headers,
@@ -38,7 +36,7 @@ class ExceptionMiddleware(StarletteExceptionMiddleware):
         )
 
         return Response(
-            content=json.dumps(connexion_response.body),
+            content=connexion_response.body,
             status_code=connexion_response.status_code,
             media_type=connexion_response.mimetype,
             headers=connexion_response.headers,
@@ -49,7 +47,7 @@ class ExceptionMiddleware(StarletteExceptionMiddleware):
         response = InternalServerError().to_problem()
 
         return Response(
-            content=json.dumps(response.body),
+            content=response.body,
             status_code=response.status_code,
             media_type=response.mimetype,
             headers=response.headers,

--- a/connexion/middleware/response_validation.py
+++ b/connexion/middleware/response_validation.py
@@ -56,8 +56,7 @@ class ResponseValidationOperation:
         """
         if mime_type.lower() not in [c.lower() for c in self._operation.produces]:
             raise NonConformingResponseHeaders(
-                reason="Invalid Response Content-type",
-                message=f"Invalid Response Content-type ({mime_type}), "
+                detail=f"Invalid Response Content-type ({mime_type}), "
                 f"expected {self._operation.produces}",
             )
 
@@ -75,9 +74,9 @@ class ResponseValidationOperation:
         if missing_keys:
             pretty_list = ", ".join(missing_keys)
             msg = (
-                "Keys in header don't match response specification. Difference: {}"
+                "Keys in response header don't match response specification. Difference: {}"
             ).format(pretty_list)
-            raise NonConformingResponseHeaders(message=msg)
+            raise NonConformingResponseHeaders(detail=msg)
 
     async def __call__(self, scope: Scope, receive: Receive, send: Send):
 

--- a/connexion/mock.py
+++ b/connexion/mock.py
@@ -41,7 +41,7 @@ class MockResolver(Resolver):
             except ResolverError as resolution_error:
                 logger.debug(
                     "... {}! Mock function is used for this operation.".format(
-                        resolution_error.reason.capitalize()
+                        resolution_error.args[0].capitalize()
                     )
                 )
                 func = mock_func

--- a/connexion/problem.py
+++ b/connexion/problem.py
@@ -4,6 +4,8 @@ This module contains a Python interface for Problem Details for HTTP APIs
 to communicate distinct "problem types" to non-human consumers.
 """
 
+import json
+
 
 def problem(status, title, detail, type=None, instance=None, headers=None, ext=None):
     """
@@ -50,5 +52,9 @@ def problem(status, title, detail, type=None, instance=None, headers=None, ext=N
     mimetype = content_type = "application/problem+json"
 
     return ConnexionResponse(
-        status, mimetype, content_type, body=problem_response, headers=headers
+        status,
+        mimetype,
+        content_type,
+        body=json.dumps(problem_response),
+        headers=headers,
     )

--- a/connexion/resolver.py
+++ b/connexion/resolver.py
@@ -70,9 +70,9 @@ class Resolver:
             return self.function_resolver(operation_id)
         except ImportError as e:
             msg = f'Cannot resolve operationId "{operation_id}"! Import error was "{str(e)}"'
-            raise ResolverError(msg, sys.exc_info())
+            raise ResolverError(msg, exc_info=sys.exc_info())
         except (AttributeError, ValueError) as e:
-            raise ResolverError(str(e), sys.exc_info())
+            raise ResolverError(str(e), exc_info=sys.exc_info())
 
 
 class RelativeResolver(Resolver):
@@ -268,9 +268,9 @@ class MethodResolverBase(RestyResolver):
             msg = 'Cannot resolve operationId "{}"! Import error was "{}"'.format(
                 operation_id, str(e)
             )
-            raise ResolverError(msg, sys.exc_info())
+            raise ResolverError(msg, exc_info=sys.exc_info())
         except (AttributeError, ValueError) as e:
-            raise ResolverError(str(e), sys.exc_info())
+            raise ResolverError(str(e), exc_info=sys.exc_info())
 
     def resolve_method_from_class(self, view_name, meth_name, view_cls):
         """

--- a/connexion/resolver.py
+++ b/connexion/resolver.py
@@ -5,7 +5,6 @@ from the operations defined in the OpenAPI spec.
 
 import inspect
 import logging
-import sys
 import typing as t
 
 from inflection import camelize
@@ -70,9 +69,9 @@ class Resolver:
             return self.function_resolver(operation_id)
         except ImportError as e:
             msg = f'Cannot resolve operationId "{operation_id}"! Import error was "{str(e)}"'
-            raise ResolverError(msg, exc_info=sys.exc_info())
+            raise ResolverError(msg)
         except (AttributeError, ValueError) as e:
-            raise ResolverError(str(e), exc_info=sys.exc_info())
+            raise ResolverError(str(e))
 
 
 class RelativeResolver(Resolver):
@@ -268,9 +267,9 @@ class MethodResolverBase(RestyResolver):
             msg = 'Cannot resolve operationId "{}"! Import error was "{}"'.format(
                 operation_id, str(e)
             )
-            raise ResolverError(msg, exc_info=sys.exc_info())
+            raise ResolverError(msg)
         except (AttributeError, ValueError) as e:
-            raise ResolverError(str(e), exc_info=sys.exc_info())
+            raise ResolverError(str(e))
 
     def resolve_method_from_class(self, view_name, meth_name, view_cls):
         """

--- a/connexion/security.py
+++ b/connexion/security.py
@@ -395,7 +395,8 @@ class SecurityHandlerFactory:
                 validation = await validation
             if not validation:
                 raise OAuthScopeProblem(
-                    detail="Provided token doesn't have the required scope",
+                    required_scopes=required_scopes,
+                    token_scopes=token_scopes,
                 )
 
             return token_info

--- a/connexion/security.py
+++ b/connexion/security.py
@@ -355,7 +355,7 @@ class SecurityHandlerFactory:
             if token_info is self.no_value:
                 return self.no_value
             if token_info is None:
-                raise OAuthResponseProblem(detail=exception_msg, token_response=None)
+                raise OAuthResponseProblem(detail=exception_msg)
             return token_info
 
         return wrapper
@@ -396,8 +396,6 @@ class SecurityHandlerFactory:
             if not validation:
                 raise OAuthScopeProblem(
                     detail="Provided token doesn't have the required scope",
-                    required_scopes=required_scopes,
-                    token_scopes=token_scopes,
                 )
 
             return token_info

--- a/connexion/validators/form_data.py
+++ b/connexion/validators/form_data.py
@@ -97,7 +97,7 @@ class FormDataValidator:
         spec_params = self.schema.get("properties", {}).keys()
         errors = set(form_params).difference(set(spec_params))
         if errors:
-            raise ExtraParameterProblem(errors, [])
+            raise ExtraParameterProblem(param_type="formData", extra_params=errors)
 
     def validate(self, data: FormData) -> None:
         if self.strict_validation:

--- a/connexion/validators/json.py
+++ b/connexion/validators/json.py
@@ -203,7 +203,7 @@ class JSONResponseBodyValidator:
         try:
             return json.loads(body)
         except json.decoder.JSONDecodeError as e:
-            raise BadRequestProblem(str(e))
+            raise NonConformingResponseBody(str(e))
 
     async def send(self, message: t.MutableMapping[str, t.Any]) -> None:
         self._messages.append(message)

--- a/connexion/validators/json.py
+++ b/connexion/validators/json.py
@@ -196,7 +196,7 @@ class JSONResponseBodyValidator:
                 extra={"validator": "body"},
             )
             raise NonConformingResponseBody(
-                message=f"{exception.message}{error_path_msg}"
+                detail=f"Response body does not conform to specification. {exception.message}{error_path_msg}"
             )
 
     def parse(self, body: str) -> dict:

--- a/connexion/validators/parameter.py
+++ b/connexion/validators/parameter.py
@@ -104,7 +104,9 @@ class ParameterValidator:
             query_errors = self.validate_query_parameter_list(request)
 
             if query_errors:
-                raise ExtraParameterProblem([], query_errors)
+                raise ExtraParameterProblem(
+                    param_type="query", extra_params=query_errors
+                )
 
         for param in self.parameters.get("query", []):
             error = self.validate_query_parameter(param, request)

--- a/tests/api/test_errors.py
+++ b/tests/api/test_errors.py
@@ -75,7 +75,7 @@ def test_errors(problem_app):
     assert problem_body["amount"] == 23.0
 
     problem_as_exception = app_client.get("/v1.0/problem_exception_with_extra_args")
-    assert problem_as_exception.status_code == 400
+    assert problem_as_exception.status_code == 500
     problem_as_exception_body = problem_as_exception.json()
     assert "age" in problem_as_exception_body
     assert problem_as_exception_body["age"] == 30

--- a/tests/api/test_headers.py
+++ b/tests/api/test_headers.py
@@ -29,10 +29,10 @@ def test_header_not_returned(simple_openapi_app):
     assert response.headers.get("content-type") == "application/problem+json"
     data = response.json()
     assert data["type"] == "about:blank"
-    assert data["title"] == "Response headers do not conform to specification"
+    assert data["title"] == "Internal Server Error"
     assert (
         data["detail"]
-        == "Keys in header don't match response specification. Difference: location"
+        == "Keys in response header don't match response specification. Difference: location"
     )
     assert data["status"] == 500
 

--- a/tests/api/test_secure_api.py
+++ b/tests/api/test_secure_api.py
@@ -105,9 +105,8 @@ def test_security(oauth_requests, secure_endpoint_app):
     assert get_bye_wrong_scope.status_code == 403
     assert get_bye_wrong_scope.headers.get("content-type") == "application/problem+json"
     get_bye_wrong_scope_reponse = get_bye_wrong_scope.json()
-    assert (
-        get_bye_wrong_scope_reponse["detail"]
-        == "Provided token doesn't have the required scope"
+    assert get_bye_wrong_scope_reponse["detail"].startswith(
+        "Provided token does not have the required scope"
     )
 
     headers = {"Authorization": "Bearer 300"}

--- a/tests/decorators/test_security.py
+++ b/tests/decorators/test_security.py
@@ -277,15 +277,15 @@ async def test_verify_security_oauthproblem():
     "errors, most_specific",
     [
         ([OAuthProblem()], OAuthProblem),
-        ([OAuthProblem(), OAuthScopeProblem([], [])], OAuthScopeProblem),
+        ([OAuthProblem(), OAuthScopeProblem()], OAuthScopeProblem),
         (
-            [OAuthProblem(), OAuthScopeProblem([], []), BadRequestProblem],
+            [OAuthProblem(), OAuthScopeProblem(), BadRequestProblem],
             OAuthScopeProblem,
         ),
         (
             [
                 OAuthProblem(),
-                OAuthScopeProblem([], []),
+                OAuthScopeProblem(),
                 BadRequestProblem,
                 ConnexionException,
             ],

--- a/tests/decorators/test_security.py
+++ b/tests/decorators/test_security.py
@@ -88,7 +88,9 @@ async def test_verify_oauth_scopes_remote(monkeypatch):
         await wrapped_func(request)
 
     assert exc_info.value.status_code == 403
-    assert exc_info.value.detail == "Provided token doesn't have the required scope"
+    assert exc_info.value.detail.startswith(
+        "Provided token does not have the required scope"
+    )
 
     tokeninfo["scope"] += " admin"
     assert await wrapped_func(request) is not None
@@ -98,7 +100,9 @@ async def test_verify_oauth_scopes_remote(monkeypatch):
         await wrapped_func(request)
 
     assert exc_info.value.status_code == 403
-    assert exc_info.value.detail == "Provided token doesn't have the required scope"
+    assert exc_info.value.detail.startswith(
+        "Provided token does not have the required scope"
+    )
 
     tokeninfo["scope"].append("admin")
     assert await wrapped_func(request) is not None
@@ -138,7 +142,9 @@ async def test_verify_oauth_scopes_local():
         await wrapped_func(request)
 
     assert exc_info.value.status_code == 403
-    assert exc_info.value.detail == "Provided token doesn't have the required scope"
+    assert exc_info.value.detail.startswith(
+        "Provided token does not have the required scope"
+    )
 
     tokeninfo["scope"] += " admin"
     assert await wrapped_func(request) is not None
@@ -148,7 +154,9 @@ async def test_verify_oauth_scopes_local():
         await wrapped_func(request)
 
     assert exc_info.value.status_code == 403
-    assert exc_info.value.detail == "Provided token doesn't have the required scope"
+    assert exc_info.value.detail.startswith(
+        "Provided token does not have the required scope"
+    )
 
     tokeninfo["scope"].append("admin")
     assert await wrapped_func(request) is not None
@@ -277,15 +285,15 @@ async def test_verify_security_oauthproblem():
     "errors, most_specific",
     [
         ([OAuthProblem()], OAuthProblem),
-        ([OAuthProblem(), OAuthScopeProblem()], OAuthScopeProblem),
+        ([OAuthProblem(), OAuthScopeProblem([], [])], OAuthScopeProblem),
         (
-            [OAuthProblem(), OAuthScopeProblem(), BadRequestProblem],
+            [OAuthProblem(), OAuthScopeProblem([], []), BadRequestProblem],
             OAuthScopeProblem,
         ),
         (
             [
                 OAuthProblem(),
-                OAuthScopeProblem(),
+                OAuthScopeProblem([], []),
                 BadRequestProblem,
                 ConnexionException,
             ],

--- a/tests/fakeapi/hello/__init__.py
+++ b/tests/fakeapi/hello/__init__.py
@@ -496,7 +496,10 @@ def get_empty_dict():
 
 def get_custom_problem_response():
     raise ProblemException(
-        403, "You need to pay", "Missing amount", ext={"amount": 23.0}
+        status=403,
+        title="You need to pay",
+        detail="Missing amount",
+        ext={"amount": 23.0},
     )
 
 

--- a/tests/test_api.py
+++ b/tests/test_api.py
@@ -5,7 +5,7 @@ from unittest.mock import MagicMock
 
 import pytest
 from connexion import FlaskApi
-from connexion.exceptions import InvalidSpecification
+from connexion.exceptions import InvalidSpecification, ResolverError
 from connexion.spec import Specification, canonical_base_path
 from yaml import YAMLError
 
@@ -60,28 +60,28 @@ def test_template():
 
 
 def test_invalid_operation_does_stop_application_to_setup():
-    with pytest.raises(ImportError):
+    with pytest.raises(ResolverError):
         FlaskApi(
             TEST_FOLDER / "fixtures/op_error_api/swagger.yaml",
             base_path="/api/v1.0",
             arguments={"title": "OK"},
         )
 
-    with pytest.raises(ValueError):
+    with pytest.raises(ResolverError):
         FlaskApi(
             TEST_FOLDER / "fixtures/missing_op_id/swagger.yaml",
             base_path="/api/v1.0",
             arguments={"title": "OK"},
         )
 
-    with pytest.raises(ImportError):
+    with pytest.raises(ResolverError):
         FlaskApi(
             TEST_FOLDER / "fixtures/module_not_implemented/swagger.yaml",
             base_path="/api/v1.0",
             arguments={"title": "OK"},
         )
 
-    with pytest.raises(ValueError):
+    with pytest.raises(ResolverError):
         FlaskApi(
             TEST_FOLDER / "fixtures/user_module_loading_error/swagger.yaml",
             base_path="/api/v1.0",

--- a/tests/test_cli.py
+++ b/tests/test_cli.py
@@ -6,6 +6,7 @@ import importlib_metadata
 import pytest
 from click.testing import CliRunner
 from connexion.cli import main
+from connexion.exceptions import ResolverError
 
 from conftest import FIXTURES_FOLDER
 
@@ -176,14 +177,14 @@ def test_run_unimplemented_operations_and_stub(mock_app_run):
     runner = CliRunner()
 
     spec_file = str(FIXTURES_FOLDER / "missing_implementation/swagger.yaml")
-    with pytest.raises(AttributeError):
+    with pytest.raises(ResolverError):
         runner.invoke(main, ["run", spec_file], catch_exceptions=False)
     # yet can be run with --stub option
     result = runner.invoke(main, ["run", spec_file, "--stub"], catch_exceptions=False)
     assert result.exit_code == 0
 
     spec_file = str(FIXTURES_FOLDER / "module_does_not_exist/swagger.yaml")
-    with pytest.raises(ImportError):
+    with pytest.raises(ResolverError):
         runner.invoke(main, ["run", spec_file], catch_exceptions=False)
     # yet can be run with --stub option
     result = runner.invoke(main, ["run", spec_file, "--stub"], catch_exceptions=False)
@@ -194,7 +195,7 @@ def test_run_unimplemented_operations_and_mock(mock_app_run):
     runner = CliRunner()
 
     spec_file = str(FIXTURES_FOLDER / "missing_implementation/swagger.yaml")
-    with pytest.raises(AttributeError):
+    with pytest.raises(ResolverError):
         runner.invoke(main, ["run", spec_file], catch_exceptions=False)
     # yet can be run with --mock option
     result = runner.invoke(

--- a/tests/test_json_validation.py
+++ b/tests/test_json_validation.py
@@ -101,7 +101,9 @@ def test_writeonly(json_validation_spec_dir, spec, app_class):
 
     res = app_client.get("/v1.0/user_with_password")
     assert res.status_code == 500
-    assert res.json()["title"] == "Response body does not conform to specification"
+    assert res.json()["detail"].startswith(
+        "Response body does not conform to specification"
+    )
 
 
 def test_nullable_default(json_validation_spec_dir, spec):

